src/runtime_src/core/common/info_aie.cpp

```diff
@@ -0,0 +1,701 @@
+/**
+ * Copyright (C) 2021 Xilinx, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may
+ * not use this file except in compliance with the License. A copy of the
+ * License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+#define XRT_CORE_COMMON_SOURCE
+#include "info_aie.h"
+#include "core/common/query_requests.h"
+#include "core/common/device.h"
+#include <boost/algorithm/string.hpp>
+#include <boost/optional/optional.hpp>
+#include <boost/property_tree/json_parser.hpp>
+
+namespace qr = xrt_core::query;
+using ptree_type = boost::property_tree::ptree;
+
+const uint32_t major = 1;
+const uint32_t minor = 0;
+const uint32_t patch = 0;
+
+namespace {
+inline std::string
+graph_status_to_string(int status) {
+  switch(status)
+  {
+    case 0:	return "stop";
+    case 1:	return "reset";
+    case 2:	return "running";
+    case 3:	return "suspend";
+    case 4:     return "end";
+    default:	return "idle";
+  }
+}
+
+// Add the node to the ouput list
+static void
+addnodelist(std::string search_str, std::string node_str, boost::property_tree::ptree& input_pt,
+	    boost::property_tree::ptree& output_pt)
+{
+  boost::property_tree::ptree pt_array;
+  for (auto& node : input_pt.get_child(search_str)) {
+    boost::property_tree::ptree pt;
+    std::string val;
+    for (auto& value : node.second)
+      val += (val.empty() ? "" : ", ") + value.second.data();
```

Comment:
At the end I do not know, especially if there is a ternary operation constructing an empty string for an identity concatenation at the end...
At least:
```c++
if (!val.empty())
  val += ", ";
val += value.second.data();
```
Otherwise there is a C++20/23 range algorithm to do this in 1 line in the future. :-)